Add -x/--line-regexp

okdana opened the pull request:
Let's try this!

This PR adds `-x`/`--line-regexp`, which essentially wraps the pattern in `^`...`$` in order to make it match whole lines only. This is an option defined in the [POSIX spec](http://pubs.opengroup.org/onlinepubs/009695399/utilities/grep.html) for `grep`, and one that i use a lot. I found it rather frustrating that `ag` didn't have it.

I will add additional notes as comments on the diff.

okdana commented on doc/convert-to-man:
```diff
@@ -1,4 +1,5 @@
-sed -i 's/\.TH.*/.TH "rg" "1"/g' rg.1
+pandoc -s -t man rg.1.md -o rg.1 &&
+sed -i.bak 's/\.TH.*/.TH "rg" "1"/g' rg.1 &&
+rm -f rg.1.bak # BSD `sed` requires the creation of a back-up file
```
I found that this script didn't run on macOS with just `sed -i`, as the BSD version of `sed` requires the back-up suffix. So i added that in a way that should be compatible with both GNU and BSD `sed`. I also added the `&&` to make the script bail if any part of the process fails.

BurntSushi replied:
> I also added the && to make the script bail if any part of the process fails.

I usually put `set -e` at the top of scripts for that. Could we do that instead? It feels a little nicer than trailing every command with `&&`. :-)

(I'm not up on my shell compatibility, but if this means switching `#!/bin/sh` to `#!/bin/bash`, then I'm fine with that.)

okdana commented on src/app.rs:
```diff
@@ -103,7 +103,8 @@ pub fn app() -> App<'static, 'static> {
              .multiple(true))
         .arg(flag("invert-match").short("v"))
-        .arg(flag("word-regexp").short("w"))
+        .arg(flag("word-regexp").short("w").overrides_with("line-regexp"))
```
It should actually be fine to apply both the `word-regexp` and `line-regexp` pattern transformations at once, since `\b` also matches the beginning/end of line. But i figured it'd be cleaner and more future-proof to make the one override the other.

okdana commented on src/args.rs:
```diff
@@ -479,10 +480,10 @@ impl<'a> ArgMatches<'a> {
     /// boundaries or escapes if applicable.
     fn str_pattern(&self, pat: &str) -> String {
-        let s = self.word_pattern(self.literal_pattern(pat.to_string()));
+        let s = self.line_pattern(self.word_pattern(self.literal_pattern(pat.to_string())));
```
This line exceeds 80 cols, and it's becoming quite unwieldy anyway, so i think maybe i should change it. I wasn't sure what would be the best way to do that though.

BurntSushi replied:
Yeah I'd definitely like to stick to a 79 column limit (inclusive). I'd probably just write it like this:

```rust
let litpat = self.literal_pattern(pat.to_string());
let s = self.line_pattern(self.word_pattern(litpat));
```

But if you want to make a new method, that's cool too!

okdana commented on src/args.rs:
```diff
@@ -511,6 +512,16 @@ impl<'a> ArgMatches<'a> {
+    /// flag is set. Otherwise, the pattern is returned unchanged.
+    fn line_pattern(&self, pat: String) -> String {
+        if self.is_present("line-regexp") {
+            format!(r"^(?:{})$", pat)
```
Added non-capturing group to avoid the same problem described in #506.

okdana commented:
btw @BurntSushi how do you feel about this? Does it fit with your vision for the tool's feature set? If so, is there anything you'd need me to do (besides rebase now) to get it in?

If not, that's cool too obv; i guess i'm just wondering if i should keep it tracking new changes or simply close it

okdana commented:
Thanks! Updated:

* Rebased
* Accounted for new zsh completion function
* Added `-e` to `convert-to-man` (it does make more sense here, and it works everywhere)
* Changed long line in `args.rs` as suggested
